### PR TITLE
[JENKINS-37011] Allow StepDescriptor to define how single argument invocation gets handled

### PR DESCRIPTION
**Pull request by kohsuke**
Allow StepDescriptor to define how single argument invocation gets handled. This is necessary to properly support Steps that do not have `@DataBoundConstructor`

### CODE REVIEW

**Comment by kohsuke** [comment]
This depends on https://github.com/jenkinsci/structs-plugin/pull/7

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java [hunk 1]
```diff
@@ -33,8 +33,16 @@
 import java.util.Set;
 
 import org.jenkinsci.plugins.structs.describable.DescribableModel;
+import org.jenkinsci.plugins.structs.describable.DescribableParameter;
+import org.kohsuke.stapler.ClassDescriptor;
+
+import javax.annotation.CheckForNull;
+
+import static com.sun.imageio.plugins.jpeg.JPEG.names;
```
Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: :ant: ...? That seems unrelated. =)

**Comment by abayer** [comment]
:bee:

**Comment by jglick** [comment]
Needs merge conflict resolution.

File: src/test/java/org/jenkinsci/plugins/workflow/steps/AbstractStepImplTest.java [hunk 2]
```diff
@@ -69,6 +72,14 @@ public void coercion() throws Exception {
         assertEquals(step.b, "pre1post");
     }
 
+    @Test
+    public void singleArgument() throws Exception {
+        assertNull(d.singleArgument("foo"));    // require two arguments
+        Map<String, Object> m = d2.singleArgument("foo");
+        assertEquals(1,m.size());
+        assertEquals("a",m.keySet().iterator().next());
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/steps/AbstractStepImplTest.java]: :ant: more concisely (and strongly):

```java
assertEquals(Collections.singletonMap("a", "foo"), d2.singleArgument("foo"));
```

**Comment by kohsuke** [comment]
After discussion with Jesse:
- [ ] prove that snippet generator works with this
- [ ] test case that covers body closure
- [ ] tests that demonstrate multiple and optional parameters

**Comment by jglick** [comment]
provisional :bee: (pending review of https://github.com/jenkinsci/workflow-cps-plugin/pull/34)

**Comment by jglick** [comment]
Does https://github.com/jenkinsci/workflow-cps-plugin/pull/59 render this obsolete?

**Comment by jglick** [comment]
Obsolete I think.
